dynamiq/nodes/agents/orchestrators/graph.py

```diff
@@ -0,0 +1,531 @@
+from dynamiq.nodes.agents.base import Agent
+from dynamiq.nodes.agents.orchestrators.graph_manager import GraphAgentManager
+from dynamiq.nodes.agents.orchestrators.orchestrator import Orchestrator, OrchestratorError
+from dynamiq.nodes.node import Node, NodeDependency
+from dynamiq.nodes.tools import Python
+from dynamiq.nodes.tools.function_tool import FunctionTool, function_tool
+from dynamiq.nodes.types import NodeGroup
+from dynamiq.runnables import RunnableConfig, RunnableStatus
+from dynamiq.utils.logger import logger
+
+
+class StateNotFoundError(OrchestratorError):
+    """Raised when next state was not found."""
+    pass
+
+
+class StateInputSchema(BaseModel):
+    context: dict[str, Any] = Field(..., description="Previous context objects.")
+    chat_history: list[dict[str, Any]] = Field(..., description="Previous chat history.")
+
+
+class State(Node):
+    """Represents single state of graph flow
+
+    Attributes:
+        name (str): Name of the state
+        description (str): Description of the state.
+        next_states (list[str]): List of adjacent node
+        tasks (list[Node]): List of tasks that have to be executed in this state.
+        condition (Python | FunctionTool): Condition that determines next state to execute.
+    """
+
+    name: str = "State"
+    group: NodeGroup = NodeGroup.UTILS
+    description: str = ""
+    next_states: list[str] = []
+    tasks: list[Node] = []
+    condition: Python | FunctionTool = None
+    manager: GraphAgentManager | None = None
+    input_schema: ClassVar[type[StateInputSchema]] = StateInputSchema
+
+    @property
+    def to_dict_exclude_params(self):
+        return super().to_dict_exclude_params | {"manager": True, "tasks": True}
+
+    def to_dict(self, **kwargs) -> dict:
+        """Converts the instance to a dictionary.
+
+        Returns:
+            dict: A dictionary representation of the instance.
+        """
+        data = super().to_dict(**kwargs)
+
+        if self.manager:
+            data["manager"] = self.manager.to_dict(**kwargs)
+        else:
+            data["manager"] = None
+        data["tasks"] = [task.to_dict(**kwargs) for task in self.tasks]
+        return data
+
+    def merge_contexts(self, context_list: list[dict[str, Any]]) -> dict:
+        """
+        Merges contexts, and raises an error when lossless merging is not possible.
+
+        Raises:
+            OrchestratorError: If multiple changes of the same variable are detected.
+        """
+        merged_dict = {}
+
+        for d in context_list:
+            for key, value in d.items():
+                if key in merged_dict:
+                    if merged_dict[key] != value:
+                        raise OrchestratorError(f"Error: multiple changes of variable {key} are detected.")
+                merged_dict[key] = value
+
+        return merged_dict
+
+    def task_description(self, task: Agent):
+        return ("Name:", task.name, "Role:", task.role)
```

Comment:
It looks a little bit weird tuple. Please use better naming, as description in most cases mean some text/string